subsys/zigbee/osif/zb_nrf_transceiver.c

```diff
@@ -257,332 +311,234 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
-		}
+	if (ack_len != ACK_PKT_LENGTH) {
+		LOG_ERR("%s: ACK length error", __func__);
+		return NET_CONTINUE;
 	}
 
-	if (rx_frame_free_slot == NULL) {
-		__ASSERT(false,
-			 "Not enough rx frames allocated for 15.4 driver");
-		return;
+	if ((*net_pkt_data(pkt) & FRAME_TYPE_MASK) != FRAME_TYPE_ACK) {
+		LOG_ERR("%s: ACK frame was expected", __func__);
+		return NET_CONTINUE;
 	}
 
-	rx_frame_free_slot->data = data;
-	rx_frame_free_slot->power = power;
-	rx_frame_free_slot->lqi = lqi;
-	rx_frame_free_slot->time = time;
+	if (ack_frame != NULL) {
+		LOG_ERR("Overwriting unhandled ACK frame.");
+	}
 
-	if (data[ACK_REQUEST_OFFSET] & ACK_REQUEST_BIT) {
-		rx_frame_free_slot->pending_bit = acked_with_pending_bit;
-	} else {
-		rx_frame_free_slot->pending_bit = ZB_FALSE;
+	ack_frame_buf[0] = ack_len;
+	if (net_pkt_read(pkt, &ack_frame_buf[1], ack_len) < 0) {
+		LOG_ERR("Failed to read ACK frame.");
+		return NET_CONTINUE;
 	}
 
-	acked_with_pending_bit = ZB_FALSE;
-	k_fifo_put(&rx_fifo, rx_frame_free_slot);
+	/* ack_frame != NULL informs that ACK frame has been received */
+	ack_frame = ack_frame_buf;
+
+	return NET_OK;
+}
+
+static enum net_verdict zigbee_l2_recv(struct net_if *iface,
+					struct net_pkt *pkt)
+{
+	ARG_UNUSED(iface);
+
+	k_fifo_put(&rx_fifo, pkt);
 
 	zb_macll_set_rx_flag();
 	zb_macll_set_trans_int();
 
-	/* Raise signal to indicate radio event */
+	/* Raise signal to indicate rx event */
 	zigbee_event_notify(ZIGBEE_EVENT_RX_DONE);
-}
 
-/* Callback notify that reception of a frame failed. */
-void nrf_802154_receive_failed(nrf_802154_rx_error_t error)
-{
-	acked_with_pending_bit = ZB_FALSE;
-
-#ifdef CONFIG_RADIO_STATISTICS
-	switch (error) {
-	case NRF_802154_RX_ERROR_NONE:
-		zb_osif_get_radio_stats()->rx_err_none++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_FRAME:
-		zb_osif_get_radio_stats()->rx_err_invalid_frame++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_FCS:
-		zb_osif_get_radio_stats()->rx_err_invalid_fcs++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_DEST_ADDR:
-		zb_osif_get_radio_stats()->rx_err_invalid_dest_addr++;
-		break;
-	case NRF_802154_RX_ERROR_RUNTIME:
-		zb_osif_get_radio_stats()->rx_err_runtime++;
-		break;
-	case NRF_802154_RX_ERROR_TIMESLOT_ENDED:
-		zb_osif_get_radio_stats()->rx_err_timeslot_ended++;
-		break;
-	case NRF_802154_RX_ERROR_ABORTED:
-		zb_osif_get_radio_stats()->rx_err_aborted++;
-		break;
-	}
-#endif /* defined CONFIG_RADIO_STATISTICS */
+	return NET_OK;
 }
 
-/* Callback notify that Energy Detection procedure finished. */
-void nrf_802154_energy_detected(uint8_t result)
+static enum net_l2_flags zigbee_l2_flags(struct net_if *iface)
 {
-	energy_detect.rssi_val = result;
-	k_sem_give(&energy_detect.sem);
-}
+	ARG_UNUSED(iface);
 
-/* Callback notifies that the energy detection procedure failed. */
-void nrf_802154_energy_detection_failed(nrf_802154_ed_error_t error)
-{
-	ARG_UNUSED(error);
-	energy_detect.failed = true;
-	k_sem_give(&energy_detect.sem);
+	return NET_L2_MULTICAST;
```

Comment:
What does it mean in Zigbee?

Reply:
I'll change it to  `return 0`.  I forgot about it after our last talk.

Reply:
fixed